File: app/src/test/java/lt/markmerkk/storage2/BasicLogStorageInsertTest.java
package lt.markmerkk.storage2;

import lt.markmerkk.DBProdExecutor;
import lt.markmerkk.storage2.database.interfaces.IQueryJob;
import org.junit.Test;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;

/**
 * Created by mariusmerkevicius on 12/13/15.
 */
public class BasicLogStorageInsertTest {
  @Test public void testValid() throws Exception {
    // Arrange
    BasicLogStorage storage = spy(new BasicLogStorage());
    storage.executor = mock(DBProdExecutor.class);
    doNothing().when(storage).notifyDataChange();

    // Act
    storage.insert(new SimpleLog());

    // Assert
    verify(storage.executor).execute(any(IQueryJob.class));
  }

  @Test public void testNullJob() throws Exception {
    // Arrange
    BasicLogStorage storage = spy(new BasicLogStorage());
    storage.executor = mock(DBProdExecutor.class);
    doNothing().when(storage).notifyDataChange();

    // Act
    storage.insert(null);

    // Assert
    verify(storage.executor, never()).execute(any(IQueryJob.class));
  }
}

File: app/src/test/java/lt/markmerkk/entities/BasicLogStorageInsertTest.kt
package lt.markmerkk.entities

import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.never
import com.nhaarman.mockito_kotlin.verify
import lt.markmerkk.entities.database.interfaces.IExecutor
import org.junit.Test

/**
 * Created by mariusmerkevicius on 12/13/15.
 */
class BasicLogStorageInsertTest {

    val executor: IExecutor = mock()
    val storage = FakeBasicLogStorage(executor)

    @Test
    fun testValid() {
        // Arrange
        // Act
        storage.insert(SimpleLog())

        // Assert
        verify(storage.executor).execute(any())
    }

    @Test
    fun testNullJob() {
        // Arrange
        // Act
        storage.insert(null)

        // Assert
        verify(storage.executor, never()).execute(any())
    }

    // Test classes

    class FakeBasicLogStorage(executor: IExecutor) : BasicLogStorage(executor) {
        override fun notifyDataChange() {
            // No notification
        }
    }

}